src/backend/optimizer/plan/setrefs.c

```diff
@@ -1081,12 +1080,7 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
 					splan->returningLists = newRL;
 
 					/*
-					 * Set up the visible plan targetlist as being the same as
-					 * the first RETURNING list. This is for the use of
-					 * EXPLAIN; the executor won't pay any attention to the
-					 * targetlist.  We postpone this step until here so that
-					 * we don't have to do set_returning_clause_references()
-					 * twice on identical targetlists.
+					 * copy the new returningList as it may be needed later.
 					 */
 					splan->plan.targetlist = copyObject(linitial(newRL));
```

Comment:
Maybe in this place find empty target lists for the material and shared scan nodes and fill them in?

Reply:
are we can travel through plan tree from bottom to up?

Reply:
Or, at a slightly later stage, go down from the specified nodes with empty target lists and fill them.

Reply:
this approach does not fix the problem since one of the Shared Scans does not have bottom nodes. look like it get a targetlist at an earlier stage.